In [2]:
# Importing the library
import pandas as pd
import numpy as np
import os


In [6]:
df_test = pd.DataFrame()

In [7]:
#create a dataframe with mixed datatype 
df_test['mix'] = ['1','2','3','4','5','6','7','8','9','10','none','a','b',True,False]

df_test.head()

mix
0   1
1   2
2   3
3   4
4   5

In [8]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
df_test.head(15)

mix
0       1
1       2
2       3
3       4
4       5
5       6
6       7
7       8
8       9
9      10
10   none
11      a
12      b
13   True
14  False

In [3]:
df_prod = pd.read_csv(r"C:\Users\okumb\Downloads\instaCart_Bucket_04_2024\Data\Original_Data\products.csv")
df_prod.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [13]:
df_prod.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
df_prod.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49677 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [16]:
df_prod.shape

(49693, 5)

In [19]:
df_pro_nan = df_prod[df_prod['product_name'].isnull() == True]
df_pro_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [4]:
df_prod_clean = df_prod[df_prod['product_name'].isnull() == False]
df_prod_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [21]:
df_prod_clean.shape

(49677, 5)

In [5]:
# getting all the duplicated value in the product_name column
df_dupl = df_prod_clean[df_prod_clean['product_name'].duplicated()]
df_dupl

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [6]:
df_prod_clean_noDupl = df_prod_clean.drop_duplicates()

In [7]:
df_prod_clean_noDupl.shape

(49672, 5)

In [29]:
# importing the order dataframe
df_order = pd.read_csv(r"C:\Users\okumb\Downloads\instaCart_Bucket_04_2024\Data\Prepared_Data\order_wrangled.csv")
df_order.head()



Unnamed: 0  order_id  user_id  order_number  order_days_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [33]:
# deleting the unnamed column in the order dataframe
df_ord = df_order.drop(columns = ['Unnamed: 0'])
df_ord.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [38]:
# checking for mixed datatype in the order dataframe

for col in df_ord.columns.tolist():
  weird = (df_ord[[col]].applymap(type) != df_ord[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ord[weird]) > 0:
    print (col)
else: print ('There is no mixed datatype in the order dataframe')


There is no mixed datatype in the order dataframe


In [36]:
# finding missing values in the order dataframe

df_ord.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
order_days_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [39]:
df_ord.shape

(3421083, 6)

### The percentage of 206,209 out of 3,421,083 is approximately 6.03%. so these empty values can be dropped without it affecting the signifcant general result of the analysis 

In [43]:
# appropiate method to deal with the missing values
df_ord.dropna(inplace = True)
df_ord.head()

order_id  user_id  order_number  order_days_of_week  order_hour_of_day  \
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   

   days_since_prior_order  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0

In [44]:
df_ord.shape

(3214874, 6)

### The difference between 3,421,083 and 3,214,874 is 206,209.


In [47]:
### getting the duplicated values in the order dataframe

df_ord_dupl = df_ord[df_ord.duplicated()]
df_ord_dupl

Empty DataFrame
Columns: [order_id, user_id, order_number, order_days_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#### from the above analysis done, there are no duplicated values in the order dataframe

In [10]:
### Exporting the order dataframe to csv file

path = r"C:\Users\okumb\Downloads\instaCart_Bucket_04_2024\Data\Prepared_Data"

In [50]:
path

'C:\\Users\\okumb\\Downloads\\instaCart_Bucket_04_2024\\Data\\Prepared_Data'

In [51]:
### Exporting the order dataframe csv file to the prepared data folder 

df_ord_dupl.to_csv(os.path.join(path, 'Order_cleaned_noDupl.csv'), index = False)


In [11]:
### Exporting the product csv file to the prepared data folder 
df_prod_clean_noDupl.to_csv(os.path.join(path, 'Product_cleaned_noDupl.csv'), index = False)

In [8]:
df_prod_clean_noDupl.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3